ignite/engine/engine.py

```diff
@@ -742,33 +747,43 @@ def _internal_run(self) -> State:
         self.should_terminate = self.should_terminate_single_epoch = False
         self._init_timers(self.state)
         try:
-            start_time = time.time()
-            self._fire_event(Events.STARTED)
-            while not self._is_done(self.state) and not self.should_terminate:
-                self.state.epoch += 1
-                self._fire_event(Events.EPOCH_STARTED)
-
-                if self._dataloader_iter is None:
-                    self._setup_engine()
-
-                time_taken = self._run_once_on_dataset()
-                # time is available for handlers but must be update after fire
-                self.state.times[Events.EPOCH_COMPLETED.name] = time_taken
-                handlers_start_time = time.time()
-                if self.should_terminate:
-                    self._fire_event(Events.TERMINATE)
-                else:
+            try:
+                start_time = time.time()
+                self._fire_event(Events.STARTED)
+                self._maybe_terminate()
+
+                while not self._is_done(self.state) and not self.should_terminate:
+                    self.state.epoch += 1
+                    handlers_start_time = time.time()
+                    self._fire_event(Events.EPOCH_STARTED)
+                    epoch_time_taken = time.time() - handlers_start_time
+                    self._maybe_terminate()
+
+                    if self._dataloader_iter is None:
+                        self._setup_engine()
+
+                    epoch_time_taken += self._run_once_on_dataset()
+
+                    # time is available for handlers but must be updated after fire
+                    self.state.times[Events.EPOCH_COMPLETED.name] = epoch_time_taken
+
+                    handlers_start_time = time.time()
                     self._fire_event(Events.EPOCH_COMPLETED)
```

Comment:
If user calls `terminate_epoch` during `_run_once_on_dataset`, Do we still want to emit `EPOCH_COMPLETED` event while run is terminated after iteration and moreover `TERMINATE_SINGLE_EPOCH` has been emitted?

Reply:
Yes, I think it is still makes sense to call EPOCH_COMPLETED even if terminate_epoch() was called.
Check this example: https://github.com/pytorch/ignite/blob/d8af03c2e34e8dff22e17570a32aedd0b755ff82/examples/reinforcement_learning/reinforce.py#L94-L116